README.md
### Disclaimer
If you are reading this please understand that repository is unfinished. This repository requires public status temporarily for testing purposes. It is still littered with errors, is unstable and the documentation is incomplete and littered with spuling errurs. Feel free to explore source code if you must but keep your commentary to yourself until I formally release. Thankyou

<img
  src="resources/temp_logo.png"
  style="display: inline-block; margin: 0 auto; width: 200px">


## Table of contents

<!--ts-->
   * [About](#about)
      * [Overview](#overview)
      * [Note From The Developer](#note-from-the-developer)
      * [Main Points](#main-points)
   * [Installation](#installation)
      * [Linux Installation](#linux-installation)
      * [MacOS Installation](#macos-installation)
      * [Windows Installation](#windows-installation)
      * [Docker Installation](#docker-installation)
   * [Getting Started](#getting-started)
   * [Generating Output](#generating-output)
      * [Generated Workspace Directory](#generating-workspace-directory)
      * [Running openscad](#running-openscad)
   * [3D Printing](#3d-printing)
      * [Running Cura - Rough Overview](#running-cura-rough-overview)
   * [Concepts](#concepts)
      * [system-codes](#system-codes)
      * [family-codes](#family-codes)
      * [type-codes](#type-codes)
      * [schemas](#schemas)
   * [Community](#community)
     * [Roadmap](#roadmap)
     * [Contributing](#contributing)
     * [FAQs](#faqs)
     * [Contacts](#contacts)
     * [Acknowledgments](#acknowledgments)
<!--te-->



# About

μpy or mupy is a shorthand for python manufacturing utility and reffers to a new system development technology implemented in the form of a python pip installable package. The μpy package essentially proposes to the open source community in the form of a powerful tool, a model for designing, distributing and manufacturing technology with an emphasis on community, modularity, decentralization, scaling, customization, parametrization, encodability and certification. 


## Overview

It is difficult to explain the full scope of this packages capabilities (both current and intended) but this overview paints a fairly simple picture of the general process a user may adopt. 

### Program some code

1. Import the mupy library

```python

import mupy.core as mu

```

2. Set up workspace directory

```python
workspace_name = "custom_box"
workspace = mu.WorkSpace(str(Path.home())+"/"+workspace_name) # Creates workspace directory relative to home path.

```

3. Defines hardware by name and system-code. 


```python
panel_a = mu.Hardware("panel_a", "CUBX0177-BPAN-B25SR2P5-X8Y8P18-RT-SX25Y25-X8Y8-X20Y20Z5")
panel_b = mu.Hardware("panel_b", "CUBX0177-BPAN-B25SR2P5-X8Y8P18-RT-SX25Y25-X8Y8-X20Y20Z5")
panel_c = mu.Hardware("panel_c", "CUBX0177-BPAN-B25SR2P5-X8Y8P18-RT-SX25Y25-X8Y8-X20Y20Z5")
panel_d = mu.Hardware("panel_d", "CUBX0177-BPAN-B25SR2P5-X8Y8P18-RT-SX25Y25-X8Y8-X20Y20Z5")
panel_e = mu.Hardware("panel_e", "CUBX0177-BPAN-B25SR2P5-X8Y8P18-RT-SX25Y25-X8Y8-X20Y20Z5")
panel_f = mu.Hardware("panel_f", "CUBX0177-BPAN-B25SR2P5-X8Y8P18-RT-SX25Y25-X8Y8-X20Y20Z5")

```

4. Define initial and final coordinates within an assembly epoch.

```python



panel_a_assembly_coords = mu.Coordinates(0,1,[0, 0 , 8 * 15 / 2 + 210],[0, 0 , 8 * 15 / 2],[0,0,0],[0,0,0])
panel_b_assembly_coords = mu.Coordinates(0,1,[0, 20 * 15 / 2 + 210, 0],[0, 20 * 15 / 2, 0],[0,0,0],[-90,0,0]) 

```

5. Defines assembly

```python
alpha = 200 
chamber_assembly = mu.Assembly("chamber_assembly")
chamber_assembly.include(panel_a, Coordinates(0,1,[0, 0 , 8 * 25 / 2 + alpha],[0, 0 , 8 * 25 / 2 ],[0,0,1000],[0,0,0]))
chamber_assembly.include(panel_b, Coordinates(0,1,[0, 8 * 25 / 2 + alpha, 0],[0, 8 * 25 / 2, 0],[0,0,1000],[-90,0,0]))
chamber_assembly.include(panel_c, Coordinates(0,1,[ - 8 * 25 / 2 - alpha, 0, 0 ],[ - 8 * 25 / 2, 0, 0 ],[0,0,1000],[0,-90,0]))
chamber_assembly.include(panel_d, Coordinates(0,1,[0 , 0, -8 * 25 / 2 - alpha],[0 , 0, -8 * 25 / 2 ],[0,0,1000],[0,180,0]))
chamber_assembly.include(panel_e, Coordinates(0,1,[0, - 8 * 25 / 2 - alpha, 0 ],[0, - 8 * 25 / 2, 0 ],[0,0,1000],[90,0,0]))
chamber_assembly.include(panel_f, Coordinates(0,1,[8 * 25 / 2 + alpha, 0 , 0],[8 * 25 / 2, 0 , 0],[0,0,1000],[90,0,90]))
```

6. Run assembly

```python
workspace.run(chamber_assembly, Coordinates(0,1,[0,0,0],[0,0,0],[0,0,0],[0,0,0]))

```

### Simulate assembly

Once this script is run, it will generate a workspace directory with generated scad source code files. These files build a hierarchical stucture for defining the overall assembly. This breaks things into steps with no specific order. These files are marked with an 'A' or a 'P' at the beginning of thier hash in the file name to signify an assembly or a hardware part respectivly. This code is used to render stl files and then import them into an assembly hierarchy which can be simulated from any point in the assembly. Just click on intended assembly .scad file.

![Alt Text](resources/utility_box.gif)

This example gives users the resources to construct a simple box. This example utilizes the CUBX0177 family and the mupy.core library. For this script, modular principles are employed however generally speaking system codes only fit with other certain system codes. In the case of the CUBX0177 family the first five elements composing the system code strings should be synchronized for modularity to be possible. Keep in mind that every type code has its own rules that it must follow and in some cases specific parameterization choices will override other intended features. In other cases the systems codes won't even be meaningful geometrically unless the libraries have built in their own boundary conditions to keep certain codes un-renderable.

### Print Parts

To print parts just import the generated stl file into [Ultimaker Cura](https://ultimaker.com/software/ultimaker-cura) and proceed with setting up your 3D-print configuration. In general all 3D-printed parts require configuration so be sure to familiarize yourself with the process. [This video runs through the process quite well](https://www.youtube.com/watch?v=T-Z3GmM20JM)

### Clean Parts
Depending on the part, you may need to clean it for it to be functionable. This is not always the case and many of the parts mupy provides do not render any wasted filiment to begin with but this is not the general case when 3D-printing.

### Assemble parts

So this is the fun part. You assemble the printed parts and make sure they fit all together. Good Luck!


## Note from the developer

Let me be the first to say that this system is rudimentary, crude, over-engineered in some aspects while under-engineered in others. It's littered with innefficiencies and design principle violations but despite all this it is extremely powerful in the right hands. The intentions behind this system are difficult to explain but in the most basic sense I needed a tool that could handle higher order levels of design, complexity, size and control in all aspects in the domain of science, engineering, manufacturing, design and distribution. It will make more sense once you use it. Good luck.

## Main Points
A collection or list of quick concepts, items and keywords which give an overview of whats mupy is all about.

### What is it?

Originally this project could of been thought of as an attempt to create a physical analogy to a software programming language ; a hardware assembly language, and it was treated as such for a time but eventually as even more abstractions and degrees of freedom were built in the hardware assembly language interpretation became not entirely appropriate. Upon further realizations about the nature about manufacturing and development set in, new concepts and tools to make eveything work. Additionally since openscad is an intimidating language due to the requirements of mathematics and programming knowledge it is not unreasonable to expect that it's popularity would not be matched by it's power so it made sense to build a system that would extend it's capability to others. It's best to think of mupy as a

- Creation Engine

- Knowledge Transfer Mechanism

- Supply Chain Mechanism

- Compression Algorithm

- Decentralized Innovation Utility

- 3D Part Generator & Library

- Applications-Programming & Command-Line Interface

- Horizontal abstraction library for openscad programming

### How do I use mupy?

There are two main interfaces for mupy, a command line interface (CLI) and an applications programming interface (API) They are described below.

- mucli - This is the manufacturing utility command line tool and is used to express the most rudimentary function of mupy ; to quickly render manufacturing resources from known system-code.

- mu.core - This library represents the general case for mupy use-case.

    - python3 scripting interface.

    - hardware class objects combine system-codes and names.

### What are mupy's functions?

- Hardware Generation

- Physical / Mechanical System Simulation

  - Operation
  - Assembly

- Physical / Mechanical System Configuration

history.md

# History

This package was developed over six year period where it began as a thought experiment which then turned into a real experiment which then became a set of research and development operations which then became a set of programmable utilities for 3D printing purposes and then that was re-envisioned an API / CLI python package built for open source.


## Context

In the beginning before I began development of mupy I had experienced a plethora of inefficiencies, difficulties and realities regarding research, development and manufacturing inefficiencies which were so difficult to overcome that I considered quitting engineering altogether. I have very big goals but not the kind you kind. My problem was my that I essentially needed to scale my developments on such a level and needed to build bigger systems (which could tap into efficiency of scale functions) but I lacked the time and resources to do so and when I finally did accomplish something interesting I seemed to lack the passion to guide the projects thenceforwards. I needed something new and something I could build momentum on for the rest of my life, in-fact it would be built into the project. A game changer really. I decided that I would only focus on a big goal and the requirements and through constraints to see that goal through. 


### Primary Requirement & Constraint : Mandate Extreme Requirements and Constraints
Mandating an extreme set of requirements and constraints will force a good idea if it exists. If the system is required to do impossible things then it will do impossible things if built. The primary constraint and goal

  - Mandate Primary Constraint : Build from nothing except what you have. 
  - Mandate Primary Goal : Build a city in space and everything we need to take us there.
  - Mandate Primary Requirement : Build a utility which will assist in that goal.

  I know this sounds crazy but it is possible and I built mupy to prove it. It is my offering. We just need to tap into efficiencies which are there and waiting to be exacted.
 

## Project ρ
Project ρ (Recursive Hardware Operations) was dedicated to making better use of smaller 3D printers by researching and developing hardware with advanced modularity features.



ask if there was some set of optimal shapes / limited parts which could make it easier to print and assemble anything. The idea was borrowed from legos except that the parts would need to be designed with specific requirements in mind that would improve the quality of any assembly especially when it comes to binding and unbinding.


### General Modularity Requirements
Imagine basic legos are provided to you along with Knex, Erector Sets and Some RoBlocks to spice the deal and imagine you were supplied enough hardware such that you could build almost anything you want say a space ship or something but you realize that the hardware is not optimized for the assembly of certain things or it could be better or something Well this was the biasses for this new approach to researching and developing modular hardware ; It was about having options to build things and a space ship was a hard test that would supply the requirements needed to make the [arts really optimal.

  - Maximize Material Independence / Material Configuration
  - Maximize Constructibility / Destructibility Freedom
  - Minimize Manufacturing Time / Cost / Resources
  - Maximize Bind-ability / Connectivity / Adaptability
  - Minimize Required Tools / Assembly Time
  - Maxmize Connective Hardware Count

### CUBX0012 Family


![Alt Text](resources/rho_legend_0.png)

The flagship family which represented project ρ was the CUBX0012 family. This cubic family propelled project ρ since CUBX0012 was so general purpose. It can be thought of as a family making up hardware elements which the simplest form attempted to build boxes from side panels ; a box with a cavity if cut up into six equal shapes would represent the general archetype geometry constituting CUBX0012 hardware elements. From more simple parts can more complex geometry be derived making up new types which belong to CUBX0012. CUBX0012 is currently retired but will be replaced soon.


### Goals

### Lessons


examples/app.py
#!/usr/bin/env python3

import mupy.core as mu


# A workspace represents a virtual 3D spce which serves as a virtual workbench except it is not a surface but a space.
""" Workspace """

workspace = WorkSpace("/home/mryan/Desktop/psi/modular_gearboxes/")

""" Hardware """


bevel_gear_a = Hardware("inner_chamber_bevel_gear_a", "CRSPGR022-BG-M1-T34W23S10-P15H0C45-B0") # Declare bevel gear of 
bevel_gear_b = Hardware("inner_chamber_bevel_gear_b", "CRSPGR022-BG-M2-T22W23S10-P15H0C45-B0")
bevel_gear_c = Hardware("inner_chamber_bevel_gear_c", "CRSPGR022-BG-M2-T34W23S10-P15H0C45-B0")
bevel_gear_d = Hardware("inner_chamber_bevel_gear_d", "CRSPGR022-BG-M3-T55W23S10-P15H0C45-B0")

bearing_a = Hardware("bearing_a", "bearing.stl")
bearing_b = Hardware("bearing_b", "bearing.stl")
bearing_c = Hardware("bearing_c", "bearing.stl")
bearing_d = Hardware("bearing_d", "bearing.stl")

panel_a = Hardware("panel_a", "CUBX0177-BPAN-B25SR2P5-X8Y8P18-RT-SX25Y25-X8Y8-X20Y20Z5")
panel_b = Hardware("panel_b", "CUBX0177-BPAN-B25SR2P5-X8Y8P18-RT-SX25Y25-X8Y8-X20Y20Z5")
panel_c = Hardware("panel_c", "CUBX0177-BPAN-B25SR2P5-X8Y8P18-RT-SX25Y25-X8Y8-X20Y20Z5")
panel_d = Hardware("panel_d", "CUBX0177-BPAN-B25SR2P5-X8Y8P18-RT-SX25Y25-X8Y8-X20Y20Z5")
panel_e = Hardware("panel_e", "CUBX0177-BPAN-B25SR2P5-X8Y8P18-RT-SX25Y25-X8Y8-X20Y20Z5")
panel_f = Hardware("panel_f", "CUBX0177-BPAN-B25SR2P5-X8Y8P18-RT-SX25Y25-X8Y8-X20Y20Z5")

flywheel_adapter_a = Hardware("flywheel_adapter_a", "CUBX0177-BPAN-B25SR2P5-X3Y7P18-RT-SX25Y25-X8Y8-X20Y2010Z5")
flywheel_adapter_b = Hardware("flywheel_adapter_b", "CUBX0177-BPAN-B25SR2P5-X3Y7P18-RT-SX25Y25-X8Y8-X20Y2010Z5")

axle_adapter_a = Hardware("axle_adapter_a", "CUBX0177-BPAN-B25SR2P5-X3Y7P18-RT-SX25Y25-X8Y8-X20Y2010Z5")
axle_adapter_b = Hardware("axle_adapter_b", "CUBX0177-BPAN-B25SR2P5-X3Y7P18-RT-SX25Y25-X8Y8-X20Y2010Z5")

axle_adapter_bolt_a_1 = Hardware("axle_adapter_bolt_a_1", "CUBX0177-BPAN-B25SR2P5-X3Y7P18-RT-SX25Y25-X8Y8-X20Y2010Z5")
axle_adapter_bolt_a_2 = Hardware("axle_adapter_bolt_a_2", "CUBX0177-BPAN-B25SR2P5-X3Y7P18-RT-SX25Y25-X8Y8-X20Y2010Z5")
axle_adapter_bolt_a_3 = Hardware("axle_adapter_bolt_a_3", "CUBX0177-BPAN-B25SR2P5-X3Y7P18-RT-SX25Y25-X8Y8-X20Y2010Z5")
axle_adapter_bolt_a_4 = Hardware("axle_adapter_bolt_a_4", "CUBX0177-BPAN-B25SR2P5-X3Y7P18-RT-SX25Y25-X8Y8-X20Y2010Z5")

axle_adapter_bolt_b_1 = Hardware("axle_adapter_bolt_b_1", "CUBX0177-BPAN-B25SR2P5-X3Y7P18-RT-SX25Y25-X8Y8-X20Y2010Z5")
axle_adapter_bolt_b_2 = Hardware("axle_adapter_bolt_b_2", "CUBX0177-BPAN-B25SR2P5-X3Y7P18-RT-SX25Y25-X8Y8-X20Y2010Z5")
axle_adapter_bolt_b_3 = Hardware("axle_adapter_bolt_b_3", "CUBX0177-BPAN-B25SR2P5-X3Y7P18-RT-SX25Y25-X8Y8-X20Y2010Z5")
axle_adapter_bolt_b_4 = Hardware("axle_adapter_bolt_b_4", "CUBX0177-BPAN-B25SR2P5-X3Y7P18-RT-SX25Y25-X8Y8-X20Y2010Z5")

boltshaft_a = Hardware("boltshaft_a", "CUBX0177-BPAN-B25SR2P5-X3Y7P18-RT-SX25Y25-X8Y8-X20Y2010Z5")
boltshaft_b = Hardware("boltshaft_a", "CUBX0177-BPAN-B25SR2P5-X3Y7P18-RT-SX25Y25-X8Y8-X20Y2010Z5")

gear_top_lock_washer_a = Hardware("gear_top_lock_washer_a","CUBX0177-BPAN-B25SR2P5-X3Y7P18-RT-SX25Y25-X8Y8-X20Y2010Z5") 
gear_top_lock_washer_b = Hardware("gear_top_lock_washer_b","CUBX0177-BPAN-B25SR2P5-X3Y7P18-RT-SX25Y25-X8Y8-X20Y2010Z5")

gear_top_washer_a = Hardware("gear_top_washer_a","CUBX0177-BPAN-B25SR2P5-X3Y7P18-RT-SX25Y25-X8Y8-X20Y2010Z5")
gear_top_washer_b = Hardware("gear_top_washer_b","CUBX0177-BPAN-B25SR2P5-X3Y7P18-RT-SX25Y25-X8Y8-X20Y2010Z5")

gear_top_nut_a = Hardware("gear_top_nut_a","CUBX0177-BPAN-B25SR2P5-X3Y7P18-RT-SX25Y25-X8Y8-X20Y2010Z5")
gear_top_nut_b = Hardware("gear_top_nut_b","CUBX0177-BPAN-B25SR2P5-X3Y7P18-RT-SX25Y25-X8Y8-X20Y2010Z5")

gear_bottom_lock_washer_a = Hardware("gear_bottom_lock_washer_a","CUBX0177-BPAN-B25SR2P5-X3Y7P18-RT-SX25Y25-X8Y8-X20Y2010Z5")
gear_bottom_lock_washer_b = Hardware("gear_bottom_lock_washer_b","CUBX0177-BPAN-B25SR2P5-X3Y7P18-RT-SX25Y25-X8Y8-X20Y2010Z5")

gear_bottom_washer_a = Hardware("gear_bottom_washer_a","CUBX0177-BPAN-B25SR2P5-X3Y7P18-RT-SX25Y25-X8Y8-X20Y2010Z5")
gear_bottom_washer_b = Hardware("gear_bottom_washer_b","CUBX0177-BPAN-B25SR2P5-X3Y7P18-RT-SX25Y25-X8Y8-X20Y2010Z5")

gear_bottom_nut_a = Hardware("gear_bottom_nut_a","CUBX0177-BPAN-B25SR2P5-X3Y7P18-RT-SX25Y25-X8Y8-X20Y2010Z5")
gear_bottom_nut_b = Hardware("gear_bottom_nut_b","CUBX0177-BPAN-B25SR2P5-X3Y7P18-RT-SX25Y25-X8Y8-X20Y2010Z5")

bearing_top_lock_washer_a = Hardware("bearing_top_lock_washer_a","CUBX0177-BPAN-B25SR2P5-X3Y7P18-RT-SX25Y25-X8Y8-X20Y2010Z5") 
bearing_top_lock_washer_b = Hardware("bearing_top_lock_washer_b","CUBX0177-BPAN-B25SR2P5-X3Y7P18-RT-SX25Y25-X8Y8-X20Y2010Z5")

bearing_top_washer_a = Hardware("bearing_top_washer_a","CUBX0177-BPAN-B25SR2P5-X3Y7P18-RT-SX25Y25-X8Y8-X20Y2010Z5")
bearing_top_washer_b = Hardware("bearing_top_washer_b","CUBX0177-BPAN-B25SR2P5-X3Y7P18-RT-SX25Y25-X8Y8-X20Y2010Z5")

bearing_top_nut_a = Hardware("bearing_top_nut_a","CUBX0177-BPAN-B25SR2P5-X3Y7P18-RT-SX25Y25-X8Y8-X20Y2010Z5")
bearing_top_nut_b = Hardware("bearing_top_nut_b","CUBX0177-BPAN-B25SR2P5-X3Y7P18-RT-SX25Y25-X8Y8-X20Y2010Z5")

bearing_bottom_lock_washer_a = Hardware("bearing_bottom_lock_washer_a","CUBX0177-BPAN-B25SR2P5-X3Y7P18-RT-SX25Y25-X8Y8-X20Y2010Z5")
bearing_bottom_lock_washer_b = Hardware("bearing_bottom_lock_washer_b","CUBX0177-BPAN-B25SR2P5-X3Y7P18-RT-SX25Y25-X8Y8-X20Y2010Z5")

bearing_bottom_washer_a = Hardware("bearing_bottom_washer_a","CUBX0177-BPAN-B25SR2P5-X3Y7P18-RT-SX25Y25-X8Y8-X20Y2010Z5")
bearing_bottom_washer_b = Hardware("bearing_bottom_washer_b","CUBX0177-BPAN-B25SR2P5-X3Y7P18-RT-SX25Y25-X8Y8-X20Y2010Z5")

bearing_bottom_nut_a = Hardware("bearing_bottom_nut_a","CUBX0177-BPAN-B25SR2P5-X3Y7P18-RT-SX25Y25-X8Y8-X20Y2010Z5")
bearing_bottom_nut_b = Hardware("bearing_bottom_nut_b","CUBX0177-BPAN-B25SR2P5-X3Y7P18-RT-SX25Y25-X8Y8-X20Y2010Z5")

bearing_adapter_top_nut_a = Hardware("bearing_bottom_nut_a","CUBX0177-BPAN-B25SR2P5-X3Y7P18-RT-SX25Y25-X8Y8-X20Y2010Z5")
bearing_adapter_top_nut_b = Hardware("bearing_bottom_nut_a","CUBX0177-BPAN-B25SR2P5-X3Y7P18-RT-SX25Y25-X8Y8-X20Y2010Z5")

bearing_adapter_top_lockwasher_a = Hardware("bearing_bottom_nut_a","CUBX0177-BPAN-B25SR2P5-X3Y7P18-RT-SX25Y25-X8Y8-X20Y2010Z5")
bearing_adapter_top_lockwasher_b = Hardware("bearing_bottom_nut_a","CUBX0177-BPAN-B25SR2P5-X3Y7P18-RT-SX25Y25-X8Y8-X20Y2010Z5")

bearing_adapter_top_washer_a = Hardware("bearing_bottom_nut_a","CUBX0177-BPAN-B25SR2P5-X3Y7P18-RT-SX25Y25-X8Y8-X20Y2010Z5")
bearing_adapter_top_washer_b = Hardware("bearing_bottom_nut_a","CUBX0177-BPAN-B25SR2P5-X3Y7P18-RT-SX25Y25-X8Y8-X20Y2010Z5")

bearing_adapter = Hardware("bearing_bottom_nut_a","CUBX0177-BPAN-B25SR2P5-X3Y7P18-RT-SX25Y25-X8Y8-X20Y2010Z5")
bearing_adapter = Hardware("bearing_bottom_nut_a","CUBX0177-BPAN-B25SR2P5-X3Y7P18-RT-SX25Y25-X8Y8-X20Y2010Z5")

bearing_adapter_bottom_washer_a = Hardware("bearing_bottom_nut_a","CUBX0177-BPAN-B25SR2P5-X3Y7P18-RT-SX25Y25-X8Y8-X20Y2010Z5")
bearing_adapter_bottom_washer_b = Hardware("bearing_bottom_nut_a","CUBX0177-BPAN-B25SR2P5-X3Y7P18-RT-SX25Y25-X8Y8-X20Y2010Z5")

bearing_adapter_bottom_lockwasher_a = Hardware("bearing_bottom_nut_a","CUBX0177-BPAN-B25SR2P5-X3Y7P18-RT-SX25Y25-X8Y8-X20Y2010Z5")
bearing_adapter_bottom_lockwasher_b = Hardware("bearing_bottom_nut_a","CUBX0177-BPAN-B25SR2P5-X3Y7P18-RT-SX25Y25-X8Y8-X20Y2010Z5")

bearing_adapter_bottom_nut_a = Hardware("bearing_bottom_nut_a","CUBX0177-BPAN-B25SR2P5-X3Y7P18-RT-SX25Y25-X8Y8-X20Y2010Z5")
bearing_adapter_bottom_nut_b = Hardware("bearing_bottom_nut_a","CUBX0177-BPAN-B25SR2P5-X3Y7P18-RT-SX25Y25-X8Y8-X20Y2010Z5")



""" Assemblies """

bearing_panel_a = Assembly("bearing_panel_a")
#bearing_panel_a.include(panel_a, Coordinates(0,1,[-999,-999,999],[10,10,10],[0,0,0],[10,10,10]))
bearing_panel_a.include(bearing_a, Coordinates(0,1,[-999,-999,999],[10,10,10],[0,0,0],[10,10,10]))

gear_axle_assembly_a = Assembly("gear_axle_assembly_a")
gear_axle_assembly_a.include(boltshaft_a, Coordinates(0,1,[-999,-999,999],[10,10,10],[0,0,0],[10,10,10]))
gear_axle_assembly_a.include(gear_top_lock_washer_a, Coordinates(0,1,[-999,-999,999],[10,10,10],[0,0,0],[10,10,10]))
gear_axle_assembly_a.include(gear_top_washer_a, Coordinates(0,1,[-999,-999,999],[10,10,10],[0,0,0],[10,10,10]))
gear_axle_assembly_a.include(bevel_gear_a, Coordinates(0,1,[-999,-999,999],[10,10,10],[0,0,0],[10,10,10]))
gear_axle_assembly_a.include(gear_bottom_washer_a, Coordinates(0,1,[-999,-999,999],[10,10,10],[0,0,0],[10,10,10]))
gear_axle_assembly_a.include(gear_bottom_lock_washer_a, Coordinates(0,1,[-999,-999,999],[10,10,10],[0,0,0],[10,10,10]))
gear_axle_assembly_a.include(gear_bottom_nut_a, Coordinates(0,1,[-999,-999,999],[10,10,10],[0,0,0],[10,10,10]))
gear_axle_assembly_a.include(bearing_top_nut_a, Coordinates(0,1,[-999,-999,999],[10,10,10],[0,0,0],[10,10,10]))
gear_axle_assembly_a.include(bearing_top_lock_washer_a, Coordinates(0,1,[-999,-999,999],[10,10,10],[0,0,0],[10,10,10]))
gear_axle_assembly_a.include(bearing_top_washer_a, Coordinates(0,1,[-999,-999,999],[10,10,10],[0,0,0],[10,10,10]))
gear_axle_assembly_a.include(bearing_panel_a, Coordinates(0,1,[-999,-999,999],[10,10,10],[0,0,0],[10,10,10]))
gear_axle_assembly_a.include(bearing_bottom_washer_a, Coordinates(0,1,[-999,-999,999],[10,10,10],[0,0,0],[10,10,10]))
gear_axle_assembly_a.include(bearing_bottom_lock_washer_a, Coordinates(0,1,[-999,-999,999],[10,10,10],[0,0,0],[10,10,10]))
gear_axle_assembly_a.include(bearing_bottom_nut_a, Coordinates(0,1,[-999,-999,999],[10,10,10],[0,0,0],[10,10,10]))
gear_axle_assembly_a.include(bearing_adapter_top_nut_a, Coordinates(0,1,[-999,-999,999],[10,10,10],[0,0,0],[10,10,10]))
gear_axle_assembly_a.include(bearing_adapter_top_lockwasher_a, Coordinates(0,1,[-999,-999,999],[10,10,10],[0,0,0],[10,10,10]))
gear_axle_assembly_a.include(bearing_adapter_top_washer_a, Coordinates(0,1,[-999,-999,999],[10,10,10],[0,0,0],[10,10,10]))
gear_axle_assembly_a.include(bearing_adapter, Coordinates(0,1,[-999,-999,999],[10,10,10],[0,0,0],[10,10,10]))
gear_axle_assembly_a.include(bearing_adapter_bottom_washer_a, Coordinates(0,1,[-999,-999,999],[10,10,10],[0,0,0],[10,10,10]))
gear_axle_assembly_a.include(bearing_adapter_bottom_lockwasher_a, Coordinates(0,1,[-999,-999,999],[10,10,10],[0,0,0],[10,10,10]))
gear_axle_assembly_a.include(bearing_adapter_bottom_nut_a, Coordinates(0,1,[-999,-999,999],[10,10,10],[0,0,0],[10,10,10]))

system_a = Assembly("system_a")
system_a.include(gear_axle_assembly_a, Coordinates(0,1,[-999,-999,999],[10,10,10],[0,0,0],[10,10,10]))
system_a.include(axle_adapter_a, Coordinates(0,1,[-999,-999,999],[10,10,10],[0,0,0],[10,10,10]))


bearing_panel_b = Assembly("bearing_panel_b")
#bearing_panel_b.include(panel_b, Coordinates(0,1,[-999,-999,999],[10,10,10],[0,0,0],[10,10,10]))
bearing_panel_b.include(bearing_b, Coordinates(0,1,[-999,-999,999],[10,10,10],[0,0,0],[10,10,10]))

gear_axle_assembly_b = Assembly("gear_axle_assembly_b")
gear_axle_assembly_b.include(boltshaft_b, Coordinates(0,1,[-999,-999,999],[10,10,10],[0,0,0],[10,10,10]))
gear_axle_assembly_b.include(gear_top_lock_washer_b, Coordinates(0,1,[-999,-999,999],[10,10,10],[0,0,0],[10,10,10]))
gear_axle_assembly_b.include(gear_top_washer_b, Coordinates(0,1,[-999,-999,999],[10,10,10],[0,0,0],[10,10,10]))
gear_axle_assembly_b.include(bevel_gear_b, Coordinates(0,1,[-999,-999,999],[10,10,10],[0,0,0],[10,10,10]))
gear_axle_assembly_b.include(gear_bottom_washer_b, Coordinates(0,1,[-999,-999,999],[10,10,10],[0,0,0],[10,10,10]))
gear_axle_assembly_b.include(gear_bottom_lock_washer_b, Coordinates(0,1,[-999,-999,999],[10,10,10],[0,0,0],[10,10,10]))
gear_axle_assembly_b.include(gear_bottom_nut_b, Coordinates(0,1,[-999,-999,999],[10,10,10],[0,0,0],[10,10,10]))
gear_axle_assembly_b.include(bearing_top_nut_b, Coordinates(0,1,[-999,-999,999],[10,10,10],[0,0,0],[10,10,10]))
gear_axle_assembly_b.include(bearing_top_lock_washer_b, Coordinates(0,1,[-999,-999,999],[10,10,10],[0,0,0],[10,10,10]))
gear_axle_assembly_b.include(bearing_top_washer_b, Coordinates(0,1,[-999,-999,999],[10,10,10],[0,0,0],[10,10,10]))
gear_axle_assembly_b.include(bearing_panel_b, Coordinates(0,1,[-999,-999,999],[10,10,10],[0,0,0],[10,10,10]))
gear_axle_assembly_b.include(bearing_bottom_washer_b, Coordinates(0,1,[-999,-999,999],[10,10,10],[0,0,0],[10,10,10]))
gear_axle_assembly_b.include(bearing_bottom_lock_washer_b, Coordinates(0,1,[-999,-999,999],[10,10,10],[0,0,0],[10,10,10]))
gear_axle_assembly_b.include(bearing_bottom_nut_b, Coordinates(0,1,[-999,-999,999],[10,10,10],[0,0,0],[10,10,10]))
gear_axle_assembly_b.include(bearing_adapter_top_nut_b, Coordinates(0,1,[-999,-999,999],[10,10,10],[0,0,0],[10,10,10]))
gear_axle_assembly_b.include(bearing_adapter_top_lockwasher_b, Coordinates(0,1,[-999,-999,999],[10,10,10],[0,0,0],[10,10,10]))
gear_axle_assembly_b.include(bearing_adapter_top_washer_b, Coordinates(0,1,[-999,-999,999],[10,10,10],[0,0,0],[10,10,10]))
gear_axle_assembly_b.include(bearing_adapter, Coordinates(0,1,[-999,-999,999],[10,10,10],[0,0,0],[10,10,10]))
gear_axle_assembly_b.include(bearing_adapter_bottom_washer_b, Coordinates(0,1,[-999,-999,999],[10,10,10],[0,0,0],[10,10,10]))
gear_axle_assembly_b.include(bearing_adapter_bottom_lockwasher_b, Coordinates(0,1,[-999,-999,999],[10,10,10],[0,0,0],[10,10,10]))
gear_axle_assembly_b.include(bearing_adapter_bottom_nut_b, Coordinates(0,1,[-999,-999,999],[10,10,10],[0,0,0],[10,10,10]))

system_b = Assembly("system_b")
system_b.include(gear_axle_assembly_b, Coordinates(0,1,[-999,-999,999],[10,10,10],[0,0,0],[10,10,10]))
system_b.include(axle_adapter_b, Coordinates(0,1,[-999,-999,999],[10,10,10],[0,0,0],[10,10,10]))

system_c = Assembly("system_c")
system_c.include(bevel_gear_b, Coordinates(0,1,[-999,-999,999],[10,10,10],[0,0,0],[10,10,10]))
system_c.include(bevel_gear_b, Coordinates(0,1,[-999,-999,999],[10,10,10],[0,0,0],[10,10,10]))

system_d = Assembly("system_d")
system_d.include(bevel_gear_b, Coordinates(0,1,[-999,-999,999],[10,10,10],[0,0,0],[10,10,10]))
system_d.include(axle_adapter_b, Coordinates(0,1,[-999,-999,999],[10,10,10],[0,0,0],[10,10,10]))


""" This is a cubic gearbox. Its should be noted that this assembly can infact be made as a class and a """

#panel_a = Part("panel_a", "CUBX0177-BPAN-B25SR2P5-X8Y8P18-RT-SX25Y25-X8Y8-X20Y20Z5")

alpha = 200 
chamber_assembly = Assembly("chamber_assembly")
chamber_assembly.include(panel_a, Coordinates(0,1,[0, 0 , 8 * 25 / 2 + alpha],[0, 0 , 8 * 25 / 2 ],[0,0,1000],[0,0,0])) # Coordinates complete!
chamber_assembly.include(panel_b, Coordinates(0,1,[0, 8 * 25 / 2 + alpha, 0],[0, 8 * 25 / 2, 0],[0,0,1000],[-90,0,0])) # Coordinates complete!
chamber_assembly.include(panel_c, Coordinates(0,1,[ - 8 * 25 / 2 - alpha, 0, 0 ],[ - 8 * 25 / 2, 0, 0 ],[0,0,1000],[0,-90,0])) # Coordinates complete!
chamber_assembly.include(panel_d, Coordinates(0,1,[0 , 0, -8 * 25 / 2 - alpha],[0 , 0, -8 * 25 / 2 ],[0,0,1000],[0,180,0])) # Coordinates complete!
chamber_assembly.include(panel_e, Coordinates(0,1,[0, - 8 * 25 / 2 - alpha, 0 ],[0, - 8 * 25 / 2, 0 ],[0,0,1000],[90,0,0])) # Coordinates complete!
chamber_assembly.include(panel_f, Coordinates(0,1,[8 * 25 / 2 + alpha, 0 , 0],[8 * 25 / 2, 0 , 0],[0,0,1000],[90,0,90])) # Coordinates complete!


workspace.run(chamber_assembly, Coordinates(0,1,[0,0,0],[0,0,0],[0,0,0],[0,0,0]))

#print(json.dumps(workspace.__dict__, indent=4, sort_keys=True))
